File: core/poziomy.py
"""
Poziomy techniczne liczone z surowych notowań — podstawa planów wejścia.

DLACZEGO TO W OGÓLE ISTNIEJE. Plany wejścia i stop-lossy stoją na KONKRETNYCH
liczbach. Model językowy poproszony o „wsparcie z wykresu" poda liczbę, która
brzmi wiarygodnie i której nikt nie jest w stanie zweryfikować. Dlatego poziomy
liczy tutaj Python, deterministycznie, a warstwa oceniająca może już tylko
WYBRAĆ jeden z policzonych — nie wymyślić własny. Bramka (`core/bramka.py`)
odrzuca każdy plan, którego poziomu nie ma na tej liście.

LICZYMY Z SUROWEGO OHLC, NIE Z MIGAWKI. To nie jest powielanie pracy skanu:
w migawkach kolumny `SMA200` i `SMA50` bywają puste dla większości spółek
(zmierzone: 1274 z 1281 w migawce 2026-09-04), a plan wejścia bez średnich
byłby ślepy. Licząc tu od zera, omijamy ten problem w całości i przy okazji
dostajemy niezależny punkt odniesienia, gdyby trzeba było go kiedyś
zdiagnozować.

KAŻDY POZIOM MA IDENTYFIKATOR I OPIS. To nie ozdoba — identyfikator jest
jedynym sposobem, w jaki plan może się powołać na poziom, a bramka to
sprawdzić.
"""
from __future__ import annotations

import pandas as pd

# Okno do wykrywania lokalnych ekstremów: punkt jest szczytem swingu, gdy jest
# najwyższy w promieniu tylu sesji. Pięć to kompromis — przy trzech łapiemy
# szum, przy dziesięciu gubimy poziomy z ostatnich dwóch tygodni.
OKNO_SWINGU = 5

# Ile ostatnich swingów każdego rodzaju zatrzymujemy. Więcej niż cztery to już
# poziomy sprzed miesięcy, których nikt nie bierze pod uwagę przy wejściu.
ILE_SWINGOW = 4

# Ile świec oddajemy do oceny. Dzienne obejmują kwartał, tygodniowe rok,
# miesięczne pięć lat — tyle wystarcza, żeby zobaczyć kontekst bez zalewania
# odbiorcy tysiącem liczb.
SWIEC_1D = 60
SWIEC_1W = 52
SWIEC_1M = 60


def _liczba(w) -> float | None:
    try:
        x = float(w)
    except (TypeError, ValueError):
        return None
    return x if pd.notna(x) else None


def atr(df: pd.DataFrame, okres: int = 14) -> float | None:
    """
    Średni rzeczywisty zakres — miara zmienności w jednostkach ceny.

    Potrzebna, bo stop-loss ma sens wyłącznie w odniesieniu do zmienności
    danego papieru. Trzy złote od kursu to przepaść przy spółce, która rusza
    się o 50 groszy dziennie, i szum przy takiej, która rusza się o pięć.
    """
    if len(df) < okres + 1:
        return None
    wysoki, niski, zamkniecie = df["High"], df["Low"], df["Close"]
    poprzednie = zamkniecie.shift(1)
    zakres = pd.concat(
        [wysoki - niski, (wysoki - poprzednie).abs(), (niski - poprzednie).abs()],
        axis=1,
    ).max(axis=1)
    wynik = _liczba(zakres.rolling(okres).mean().iloc[-1])
    return round(wynik, 4) if wynik else None


def swingi(df: pd.DataFrame, okno: int = OKNO_SWINGU) -> tuple[list, list]:
    """
    Lokalne szczyty i dołki. Zwraca listy (data, cena), od najnowszych.

    Szczyt swingu to sesja, której maksimum jest najwyższe w promieniu `okno`
    sesji w obie strony. Ostatnie `okno` sesji jest z definicji niepełne —
    nie wiemy jeszcze, czy nie padnie tam wyższy szczyt — więc ich nie
    zgłaszamy. To celowe: poziom, który może się jeszcze zmienić, nie nadaje
    się na stop-loss.
    """
    if len(df) < okno * 2 + 1:
        return [], []

    gory: list[tuple[str, float]] = []
    doly: list[tuple[str, float]] = []
    wysoki = df["High"].to_numpy()
    niski = df["Low"].to_numpy()

    for i in range(okno, len(df) - okno):
        okolica_g = wysoki[i - okno : i + okno + 1]
        okolica_d = niski[i - okno : i + okno + 1]
        dzien = df.index[i]
        etykieta = pd.Timestamp(dzien).date().isoformat()
        if wysoki[i] == okolica_g.max():
            gory.append((etykieta, round(float(wysoki[i]), 4)))
        if niski[i] == okolica_d.min():
            doly.append((etykieta, round(float(niski[i]), 4)))

    return gory[::-1], doly[::-1]


def _srednia(df: pd.DataFrame, okres: int) -> float | None:
    if len(df) < okres:
        return None
    # dropna PRZED liczeniem: pojedyncza dziura w notowaniach nie może
    # wyzerować całego okna kroczącego.
    czyste = df["Close"].dropna()
    if len(czyste) < okres:
        return None
    return round(float(czyste.rolling(okres).mean().iloc[-1]), 4)


def _trend(df: pd.DataFrame, okres: int, wstecz: int) -> str:
    """
    Kierunek trendu: cena wobec średniej PLUS nachylenie samej średniej.

    Sama relacja ceny do średniej nie wystarcza — cena bywa nad opadającą
    średnią w trakcie odbicia w trendzie spadkowym. Dopiero oba warunki naraz
    pozwalają nazwać rzecz trendem, a nie chwilowym wychyleniem.
    """
    czyste = df["Close"].dropna()
    if len(czyste) < okres + wstecz:
        return "nieokreślony"
    srednia = czyste.rolling(okres).mean()
    teraz = _liczba(srednia.iloc[-1])
    kiedys = _liczba(srednia.iloc[-1 - wstecz])
    cena = _liczba(czyste.iloc[-1])
    if teraz is None or kiedys is None or cena is None:
        return "nieokreślony"
    rosnie = teraz > kiedys
    nad = cena > teraz
    if nad and rosnie:
        return "wzrostowy"
    if not nad and not rosnie:
        return "spadkowy"
    return "boczny"


def luki(df: pd.DataFrame, ile: int = 3) -> list[dict]:
    """
    Niedomknięte luki cenowe z ostatnich sesji.

    Luka to poziom, do którego rynek często wraca, więc bywa naturalnym celem
    albo miejscem odbicia. Za domkniętą uznajemy taką, przez którą cena już
    przeszła — te pomijamy, bo nie są już żadnym poziomem.
    """
    if len(df) < 2:
        return []
    wynik: list[dict] = []
    wysoki = df["High"].to_numpy()
    niski = df["Low"].to_numpy()
    ostatnia = float(df["Close"].dropna().iloc[-1])

    for i in range(len(df) - 1, max(0, len(df) - 120), -1):
        gora_w_gore = niski[i] > wysoki[i - 1]
        gora_w_dol = wysoki[i] < niski[i - 1]
        if not (gora_w_gore or gora_w_dol):
            continue
        od = float(wysoki[i - 1]) if gora_w_gore else float(wysoki[i])
        do = float(niski[i]) if gora_w_gore else float(niski[i - 1])
        # Domknięta, gdy kurs zdążył wejść w jej zakres.
        if min(od, do) <= ostatnia <= max(od, do):
            continue
        wynik.append({
            "data": pd.Timestamp(df.index[i]).date().isoformat(),
            "od": round(min(od, do), 4),
            "do": round(max(od, do), 4),
            "kierunek": "w górę" if gora_w_gore else "w dół",
        })
        if len(wynik) >= ile:
            break
    return wynik


def _swiece(df: pd.DataFrame, ile: int) -> list[dict]:
    """Ostatnie świece w zwięzłej formie — do oceny kontekstu, nie do liczenia."""
    ogon = df.tail(ile)
    return [
        {
            "d": pd.Timestamp(i).date().isoformat(),
            "o": round(float(r["Open"]), 4),
            "h": round(float(r["High"]), 4),
            "l": round(float(r["Low"]), 4),
            "c": round(float(r["Close"]), 4),
        }
        for i, r in ogon.iterrows()
        if pd.notna(r["Close"])
    ]


def _przelicz(df: pd.DataFrame, reguła: str) -> pd.DataFrame:
    """Przeliczenie świec dziennych na tygodniowe albo miesięczne."""
    return (
        df.resample(reguła)
        .agg({"Open": "first", "High": "max", "Low": "min",
              "Close": "last", "Volume": "sum"})
        .dropna(subset=["Close"])
    )


def zbuduj(df: pd.DataFrame, kurs: float | None = None) -> dict | None:
    """
    Komplet poziomów i kontekstu dla jednej spółki.

    `df` to dzienne OHLC (indeks czasowy). Tygodniowe i miesięczne
    przeliczamy sami, zamiast pobierać osobno — mniej zapytań i pewność, że
    wszystkie trzy interwały opisują dokładnie ten sam zakres dat.
    """
    if df is None or df.empty or len(df) < 60:
        return None
    if isinstance(df.columns, pd.MultiIndex):
        df = df.copy()
        df.columns = df.columns.get_level_values(0)
    if getattr(df.index, "tz", None) is not None:
        df = df.tz_localize(None)

    zamkniecia = df["Close"].dropna()
    if zamkniecia.empty:
        return None
    cena = _liczba(kurs) or float(zamkniecia.iloc[-1])
    if cena <= 0:
        return None

    tyg = _przelicz(df, "W")
    mies = _przelicz(df, "ME")

    gory, doly = swingi(df)
    a = atr(df)
    rok = df.tail(252)
    szczyt52 = _liczba(rok["High"].max())
    dolek52 = _liczba(rok["Low"].min())

    poziomy: list[dict] = []

    def dodaj(ident: str, wartosc, opis: str, rodzaj: str) -> None:
        w = _liczba(wartosc)
        if w is None or w <= 0:
            return
        poziomy.append({
            "id": ident,
            "wartosc": round(w, 4),
            "opis": opis,
            "rodzaj": rodzaj,
            "dystans_pct": round((w - cena) / cena * 100, 2),
        })

    for nr, (dzien, w) in enumerate(doly[:ILE_SWINGOW], start=1):
        dodaj(f"swing_low_{nr}", w, f"dołek swingu 1D z {dzien}", "wsparcie")
    for nr, (dzien, w) in enumerate(gory[:ILE_SWINGOW], start=1):
        dodaj(f"swing_high_{nr}", w, f"szczyt swingu 1D z {dzien}", "opór")

    for okres in (20, 50, 200):
        dodaj(f"sma{okres}", _srednia(df, okres), f"średnia {okres}-sesyjna",
              "średnia")
    dodaj("sma_tyg_10", _srednia(tyg, 10), "średnia 10-tygodniowa", "średnia")

    dodaj("szczyt_52t", szczyt52, "szczyt 52 tygodni", "opór")
    dodaj("dolek_52t", dolek52, "dołek 52 tygodni", "wsparcie")

    for nr, l in enumerate(luki(df), start=1):
        dodaj(f"luka_{nr}_od", l["od"], f"luka {l['kierunek']} z {l['data']} (brzeg)",
              "luka")
        dodaj(f"luka_{nr}_do", l["do"], f"luka {l['kierunek']} z {l['data']} (brzeg)",
              "luka")

    wolumen = None
    if "Volume" in df.columns and len(df) >= 20:
        ostatni = _liczba(df["Volume"].iloc[-1])
        srednia20 = _liczba(df["Volume"].tail(20).mean())
        if ostatni is not None and srednia20:
            wolumen = {
                "ostatni": int(ostatni),
                "srednia_20": int(srednia20),
                "krotnosc": round(ostatni / srednia20, 2),
            }

    pozycja = None
    if szczyt52 and dolek52 and szczyt52 > dolek52:
        pozycja = round((cena - dolek52) / (szczyt52 - dolek52) * 100, 1)

    return {
        "kurs": round(cena, 4),
        "atr": a,
        "atr_pct": round(a / cena * 100, 2) if a else None,
        "trend": {
            "1d": _trend(df, 20, 10),
            "1w": _trend(tyg, 10, 4),
            "1m": _trend(mies, 12, 3),
        },
        "zakres_52t": {
            "min": szczyt52 and dolek52 and round(dolek52, 4),
            "maks": szczyt52 and round(szczyt52, 4),
            "pozycja_pct": pozycja,
        },
        "wolumen": wolumen,
        "luki": luki(df),
        "poziomy": sorted(poziomy, key=lambda p: p["wartosc"]),
        "swiece": {
            "1d": _swiece(df, SWIEC_1D),
            "1w": _swiece(tyg, SWIEC_1W),
            "1m": _swiece(mies, SWIEC_1M),
        },
        "sesji_w_historii": len(df),
    }

File: frontend/lib/poziomy.ts
/**
 * Poziomy techniczne liczone z surowego OHLC — LUSTRO `core/poziomy.py`.
 *
 * DLACZEGO ISTNIEJE DRUGA IMPLEMENTACJA TEGO SAMEGO. Pythonowa wersja liczy
 * poziomy podczas porannego dossier, na komputerze użytkownika albo w GitHub
 * Actions. Maszyna researchowa w Claude ma działać Z PRACY, dla DOWOLNEJ
 * spółki, o dowolnej porze — także takiej, której nie ma w dzisiejszym dossier
 * (dossier obejmuje ~20 kandydatów, uniwersum ma ~1300 instrumentów). Bez tej
 * kopii jedyny plan wejścia, jaki dałoby się ułożyć zdalnie, opierałby się na
 * poziomach zmyślonych przez model — czyli dokładnie na tym, czego bramka
 * w `core/bramka.py` zabrania.
 *
 * ZASADA LUSTRA: ten plik nie wymyśla własnych reguł. Każda stała, każdy próg
 * i każde zaokrąglenie pochodzi z `core/poziomy.py`. Zmiana po jednej stronie
 * wymaga zmiany po drugiej, a `scripts/porownaj_poziomy.py` sprawdza zgodność
 * na żywych danych.
 *
 * CENY ZOSTAJĄ W WALUCIE NOTOWANIA, bez przeliczania. Yahoo podaje Londyn
 * w PENSACH ("GBp"), tak samo jak migawka ze skanu, więc obie strony mówią
 * tym samym językiem. Przeliczenie tutaj rozjechałoby je i dałoby poziomy
 * 100x nie takie. Fakt notowania w subjednostkach wychodzi na wierzch jako
 * `waluta_w_podjednostkach` i ma tam zostać.
 */
import { notowania, skalaCeny } from "./transakcje";

/** Okno wykrywania ekstremów: punkt jest swingiem, gdy dominuje w tym promieniu. */
const OKNO_SWINGU = 5;
/** Ile ostatnich swingów każdego rodzaju zatrzymujemy. */
const ILE_SWINGOW = 4;
const SWIEC_1D = 60;
const SWIEC_1W = 52;
const SWIEC_1M = 60;

export type Swieca = {
  d: string;
  o: number;
  h: number;
  l: number;
  c: number;
  v: number;
};

export type Poziom = {
  id: string;
  wartosc: number;
  opis: string;
  rodzaj: "wsparcie" | "opór" | "średnia" | "luka";
  dystans_pct: number;
};

export type Luka = {
  data: string;
  od: number;
  do: number;
  kierunek: "w górę" | "w dół";
};

export type Poziomy = {
  kurs: number;
  atr: number | null;
  atr_pct: number | null;
  trend: { "1d": string; "1w": string; "1m": string };
  zakres_52t: {
    min: number | null;
    maks: number | null;
    pozycja_pct: number | null;
  };
  wolumen: { ostatni: number; srednia_20: number; krotnosc: number } | null;
  luki: Luka[];
  poziomy: Poziom[];
  swiece: { "1d": Swieca[]; "1w": Swieca[]; "1m": Swieca[] };
  sesji_w_historii: number;
};

/**
 * Zaokrąglenie zgodne z `round()` w Pythonie.
 *
 * NIE `Math.round(x * 100) / 100` — mnożenie wprowadza własny błąd
 * (0,595 x 100 daje 59,50000000000001), przez co ta sama liczba wychodzi
 * inaczej po obu stronach. Ta pułapka kosztowała już jeden rozjazd
 * statystyki planów (+0,60 R kontra +0,59 R).
 */
function zaokr(x: number, cyfry: number): number {
  return Number(x.toFixed(cyfry));
}

function skonczona(x: unknown): number | null {
  return typeof x === "number" && Number.isFinite(x) ? x : null;
}

// ---------------------------------------------------------------------------
// Przeliczanie interwałów
// ---------------------------------------------------------------------------

/** Niedziela kończąca tydzień danej daty — etykieta jak w `resample("W")`. */
function koniecTygodnia(dzien: string): string {
  const d = new Date(`${dzien}T00:00:00Z`);
  // getUTCDay(): 0 = niedziela. Do niedzieli zostaje (7 - dzień) % 7 dni.
  const doNiedzieli = (7 - d.getUTCDay()) % 7;
  d.setUTCDate(d.getUTCDate() + doNiedzieli);
  return d.toISOString().slice(0, 10);
}

/** Ostatni dzień miesiąca danej daty — etykieta jak w `resample("ME")`. */
function koniecMiesiaca(dzien: string): string {
  const d = new Date(`${dzien}T00:00:00Z`);
  // Dzień 0 następnego miesiąca to ostatni dzień bieżącego.
  const ostatni = new Date(Date.UTC(d.getUTCFullYear(), d.getUTCMonth() + 1, 0));
  return ostatni.toISOString().slice(0, 10);
}

/**
 * Świece dzienne na tygodniowe albo miesięczne: otwarcie pierwsze, maksimum
 * najwyższe, minimum najniższe, zamknięcie ostatnie, wolumen zsumowany.
 */
export function przelicz(swiece: Swieca[], interwal: "W" | "M"): Swieca[] {
  const etykieta = interwal === "W" ? koniecTygodnia : koniecMiesiaca;
  const grupy = new Map<string, Swieca>();
  for (const s of swiece) {
    const klucz = etykieta(s.d);
    const biezaca = grupy.get(klucz);
    if (!biezaca) {
      grupy.set(klucz, { d: klucz, o: s.o, h: s.h, l: s.l, c: s.c, v: s.v });
      continue;
    }
    biezaca.h = Math.max(biezaca.h, s.h);
    biezaca.l = Math.min(biezaca.l, s.l);
    biezaca.c = s.c;
    biezaca.v += s.v;
  }
  return [...grupy.values()].sort((a, b) => a.d.localeCompare(b.d));
}

// ---------------------------------------------------------------------------
// Wskaźniki
// ---------------------------------------------------------------------------

/**
 * Średni rzeczywisty zakres — zmienność w jednostkach ceny.
 *
 * Stop-loss ma sens wyłącznie wobec zmienności papieru: trzy złote to przepaść
 * przy spółce ruszającej się o 50 groszy dziennie i szum przy takiej, która
 * rusza się o pięć.
 */
export function atr(swiece: Swieca[], okres = 14): number | null {
  if (swiece.length < okres + 1) return null;
  const zakresy: number[] = [];
  for (let i = 1; i < swiece.length; i++) {
    const poprzednie = swiece[i - 1].c;
    zakresy.push(
      Math.max(
        swiece[i].h - swiece[i].l,
        Math.abs(swiece[i].h - poprzednie),
        Math.abs(swiece[i].l - poprzednie),
      ),
    );
  }
  const okno = zakresy.slice(-okres);
  if (okno.length < okres) return null;
  const wynik = okno.reduce((a, b) => a + b, 0) / okres;
  return wynik ? zaokr(wynik, 4) : null;
}

/**
 * Lokalne szczyty i dołki, od najnowszych.
 *
 * Ostatnie `okno` sesji jest z definicji niepełne — nie wiadomo jeszcze, czy
 * nie padnie tam wyższy szczyt — więc ich nie zgłaszamy. Poziom, który może
 * się jeszcze zmienić, nie nadaje się na stop-loss.
 */
export function swingi(
  swiece: Swieca[],
  okno = OKNO_SWINGU,
): { gory: [string, number][]; doly: [string, number][] } {
  if (swiece.length < okno * 2 + 1) return { gory: [], doly: [] };
  const gory: [string, number][] = [];
  const doly: [string, number][] = [];
  for (let i = okno; i < swiece.length - okno; i++) {
    const okolica = swiece.slice(i - okno, i + okno + 1);
    const maks = Math.max(...okolica.map((s) => s.h));
    const min = Math.min(...okolica.map((s) => s.l));
    if (swiece[i].h === maks) gory.push([swiece[i].d, zaokr(swiece[i].h, 4)]);
    if (swiece[i].l === min) doly.push([swiece[i].d, zaokr(swiece[i].l, 4)]);
  }
  return { gory: gory.reverse(), doly: doly.reverse() };
}

function srednia(swiece: Swieca[], okres: number): number | null {
  if (swiece.length < okres) return null;
  const zamkniecia = swiece.map((s) => s.c).filter((c) => Number.isFinite(c));
  if (zamkniecia.length < okres) return null;
  const okno = zamkniecia.slice(-okres);
  return zaokr(okno.reduce((a, b) => a + b, 0) / okres, 4);
}

/**
 * Kierunek trendu: cena wobec średniej PLUS nachylenie samej średniej.
 *
 * Sama relacja ceny do średniej nie wystarcza — cena bywa nad opadającą
 * średnią w trakcie odbicia w trendzie spadkowym.
 */
function trend(swiece: Swieca[], okres: number, wstecz: number): string {
  const zamkniecia = swiece.map((s) => s.c).filter((c) => Number.isFinite(c));
  if (zamkniecia.length < okres + wstecz) return "nieokreślony";
  const sredniaDo = (koniec: number): number | null => {
    const okno = zamkniecia.slice(koniec - okres, koniec);
    if (okno.length < okres) return null;
    return okno.reduce((a, b) => a + b, 0) / okres;
  };
  const teraz = sredniaDo(zamkniecia.length);
  const kiedys = sredniaDo(zamkniecia.length - wstecz);
  const cena = zamkniecia[zamkniecia.length - 1];
  if (teraz === null || kiedys === null || !Number.isFinite(cena)) {
    return "nieokreślony";
  }
  const rosnie = teraz > kiedys;
  const nad = cena > teraz;
  if (nad && rosnie) return "wzrostowy";
  if (!nad && !rosnie) return "spadkowy";
  return "boczny";
}

/**
 * Niedomknięte luki cenowe. Za domkniętą uznajemy taką, przez którą cena już
 * przeszła — te pomijamy, bo nie są już żadnym poziomem.
 */
export function luki(swiece: Swieca[], ile = 3): Luka[] {
  if (swiece.length < 2) return [];
  const wynik: Luka[] = [];
  const zamkniecia = swiece.map((s) => s.c).filter((c) => Number.isFinite(c));
  const ostatnia = zamkniecia[zamkniecia.length - 1];
  const granica = Math.max(0, swiece.length - 120);
  for (let i = swiece.length - 1; i > granica; i--) {
    const wGore = swiece[i].l > swiece[i - 1].h;
    const wDol = swiece[i].h < swiece[i - 1].l;
    if (!wGore && !wDol) continue;
    const od = wGore ? swiece[i - 1].h : swiece[i].h;
    const doKad = wGore ? swiece[i].l : swiece[i - 1].l;
    if (Math.min(od, doKad) <= ostatnia && ostatnia <= Math.max(od, doKad)) {
      continue;
    }
    wynik.push({
      data: swiece[i].d,
      od: zaokr(Math.min(od, doKad), 4),
      do: zaokr(Math.max(od, doKad), 4),
      kierunek: wGore ? "w górę" : "w dół",
    });
    if (wynik.length >= ile) break;
  }
  return wynik;
}

function ogon(swiece: Swieca[], ile: number): Swieca[] {
  return swiece.slice(-ile).map((s) => ({
    d: s.d,
    o: zaokr(s.o, 4),
    h: zaokr(s.h, 4),
    l: zaokr(s.l, 4),
    c: zaokr(s.c, 4),
    v: Math.round(s.v),
  }));
}

// ---------------------------------------------------------------------------
// Komplet
// ---------------------------------------------------------------------------

/**
 * Komplet poziomów i kontekstu dla jednej spółki.
 *
 * `swiece` to dzienne OHLC rosnąco. Tygodniowe i miesięczne przeliczamy sami,
 * zamiast pobierać osobno — mniej zapytań i pewność, że wszystkie trzy
 * interwały opisują dokładnie ten sam zakres dat.
 */
export function zbuduj(swiece: Swieca[], kurs?: number | null): Poziomy | null {
  if (!swiece || swiece.length < 60) return null;
  const zamkniecia = swiece.map((s) => s.c).filter((c) => Number.isFinite(c));
  if (zamkniecia.length === 0) return null;
  const cena = skonczona(kurs) ?? zamkniecia[zamkniecia.length - 1];
  if (cena <= 0) return null;

  const tyg = przelicz(swiece, "W");
  const mies = przelicz(swiece, "M");
  const { gory, doly } = swingi(swiece);
  const a = atr(swiece);

  const rok = swiece.slice(-252);
  const szczyt52 = rok.length ? Math.max(...rok.map((s) => s.h)) : null;
  const dolek52 = rok.length ? Math.min(...rok.map((s) => s.l)) : null;

  const poziomy: Poziom[] = [];
  const dodaj = (
    id: string,
    wartosc: number | null,
    opis: string,
    rodzaj: Poziom["rodzaj"],
  ): void => {
    const w = skonczona(wartosc);
    if (w === null || w <= 0) return;
    poziomy.push({
      id,
      wartosc: zaokr(w, 4),
      opis,
      rodzaj,
      dystans_pct: zaokr(((w - cena) / cena) * 100, 2),
    });
  };

  doly.slice(0, ILE_SWINGOW).forEach(([dzien, w], nr) => {
    dodaj(`swing_low_${nr + 1}`, w, `dołek swingu 1D z ${dzien}`, "wsparcie");
  });
  gory.slice(0, ILE_SWINGOW).forEach(([dzien, w], nr) => {
    dodaj(`swing_high_${nr + 1}`, w, `szczyt swingu 1D z ${dzien}`, "opór");
  });

  for (const okres of [20, 50, 200]) {
    dodaj(
      `sma${okres}`,
      srednia(swiece, okres),
      `średnia ${okres}-sesyjna`,
      "średnia",
    );
  }
  dodaj("sma_tyg_10", srednia(tyg, 10), "średnia 10-tygodniowa", "średnia");

  dodaj("szczyt_52t", szczyt52, "szczyt 52 tygodni", "opór");
  dodaj("dolek_52t", dolek52, "dołek 52 tygodni", "wsparcie");

  const listaLuk = luki(swiece);
  listaLuk.forEach((l, nr) => {
    dodaj(`luka_${nr + 1}_od`, l.od, `luka ${l.kierunek} z ${l.data} (brzeg)`, "luka");
    dodaj(`luka_${nr + 1}_do`, l.do, `luka ${l.kierunek} z ${l.data} (brzeg)`, "luka");
  });

  let wolumen: Poziomy["wolumen"] = null;
  if (swiece.length >= 20) {
    const ostatni = swiece[swiece.length - 1].v;
    const ostatnie20 = swiece.slice(-20).map((s) => s.v);
    const srednia20 =
      ostatnie20.reduce((x, y) => x + y, 0) / ostatnie20.length;
    if (Number.isFinite(ostatni) && srednia20) {
      wolumen = {
        ostatni: Math.trunc(ostatni),
        srednia_20: Math.trunc(srednia20),
        krotnosc: zaokr(ostatni / srednia20, 2),
      };
    }
  }

  const pozycja =
    szczyt52 !== null && dolek52 !== null && szczyt52 > dolek52
      ? zaokr(((cena - dolek52) / (szczyt52 - dolek52)) * 100, 1)
      : null;

  return {
    kurs: zaokr(cena, 4),
    atr: a,
    atr_pct: a ? zaokr((a / cena) * 100, 2) : null,
    trend: {
      "1d": trend(swiece, 20, 10),
      "1w": trend(tyg, 10, 4),
      "1m": trend(mies, 12, 3),
    },
    zakres_52t: {
      min: dolek52 !== null ? zaokr(dolek52, 4) : null,
      maks: szczyt52 !== null ? zaokr(szczyt52, 4) : null,
      pozycja_pct: pozycja,
    },
    wolumen,
    luki: listaLuk,
    poziomy: poziomy.sort((x, y) => x.wartosc - y.wartosc),
    swiece: {
      "1d": ogon(swiece, SWIEC_1D),
      "1w": ogon(tyg, SWIEC_1W),
      "1m": ogon(mies, SWIEC_1M),
    },
    sesji_w_historii: swiece.length,
  };
}

/**
 * Poziomy dla tickera prosto z Yahoo. `null`, gdy notowań nie ma albo jest
 * ich za mało (mniej niż 60 sesji — świeże IPO, wycofany instrument).
 */
export async function dlaTickera(
  ticker: string,
): Promise<(Poziomy & { waluta: string; waluta_w_podjednostkach: boolean }) | null> {
  // 10 lat i ceny skorygowane — DOKŁADNIE to, co pobiera
  // `scripts/przygotuj_dossier.py` (`period="10y"`, `auto_adjust=True`).
  // Krótszy okres dałby inne świece miesięczne i inny trend 1M, a ceny
  // nieskorygowane — inne swingi i inny zakres 52 tygodni na spółkach
  // dywidendowych. Lustro musi brać te same dane, nie podobne.
  const n = await notowania(ticker, "10y", true);
  if (!n) return null;
  const swiece: Swieca[] = n.dni.map((d, i) => ({
    d,
    o: n.otwarcie[i],
    h: n.najwyzsze[i],
    l: n.najnizsze[i],
    c: n.zamkniecie[i],
    v: n.wolumen[i],
  }));
  const p = zbuduj(swiece);
  if (!p) return null;
  return {
    ...p,
    waluta: n.waluta,
    waluta_w_podjednostkach: skalaCeny(n.waluta) !== 1,
  };
}
